### ISSUE TITLE
Bernoulli's probability mass function produces output beyond its support

### ISSUE THREAD

**Issue by dustinvtran** [tensorflow/tensorflow#8583]
Bernoulli's (log) probability mass function is valid for values between 0 and 1, even though its support is only {0, 1}. This is a result of [its implementation](https://github.com/tensorflow/tensorflow/blob/r1.0/tensorflow/contrib/distributions/python/ops/bernoulli.py#L150) relying on `tf.sigmoid_cross_entropy_with_logits(logits, targets)`, where `targets` does not have to be {0, 1}.

Is this intended behavior?

```python
from tensorflow.contrib import distributions as ds

x = ds.Bernoulli(p=0.5)
y = x.log_prob(0.5)

sess = tf.Session()
sess.run(y)
## -0.69314718
```

```python
print(tf.__version__)
## 1.0.1
```

**Comment by skye** [comment]
@ebrevdo @jvdillon @langmore

**Comment by ebrevdo** [comment]
This is indeed a bug.  Solution is:

1. if the value can be checked statically, we should ensure its either 0 or
1.
2. if not, and validate_args is true, we need to add a dynamic assertion.
